lib/safari.js

```diff
@@ -1,53 +1,173 @@
-import { fs, process, logger } from 'appium-support';
+import os from 'os';
+import path from 'path';
+import { JWProxy, errors } from 'appium-base-driver';
+import { fs, logger, util } from 'appium-support';
 import { SubProcess } from 'teen_process';
+import { waitForCondition } from 'asyncbox';
+import { findAPortNotInUse } from 'portscanner';
+import { execSync } from 'child_process';
 
-const log = logger.getLogger('SafariDriverProc');
+const log = logger.getLogger('SafariDriverServer');
 
-class SafariDriverProc {
-  constructor ({port, unexpectedExitHandler}) {
-    this.port = port;
+const SD_BINARY = 'safaridriver';
+const STARTUP_TIMEOUT = 10000; // seconds
+const SAFARI_PORT_RANGE = [5100, 5200];
+// This guard is needed to make sure
+// we never run multiple Safari driver processes for the same Appium process
+const SAFARI_SERVER_GUARD = util.getLockFileGuard(
+  path.resolve(os.tmpdir(), 'safari_server_guard.lock'),
+  {timeout: 5, tryRecovery: true}
+);
+
+
+class SafariProxy extends JWProxy {
+  async proxyCommand (url, method, body = null) {
+    if (this.didProcessExit) {
+      throw new errors.InvalidContextError(
+        `'${method} ${url}' cannot be proxied to Safari Driver server because ` +
+        'the process is not running (probably crashed). Check the server log for more details');
+    }
+    return await super.proxyCommand(url, method, body);
+  }
+}
+
+
+class SafariDriverProcess {
+  constructor () {
+    this.port = null;
     this.proc = null;
-    this.unexpectedExitHandler = unexpectedExitHandler;
   }
 
-  async start () {
-    await process.killProcess('safaridriver');
-    const safariBin = await fs.which('safaridriver');
-    this.proc = new SubProcess(safariBin, ['-p', this.port]);
+  get isRunning () {
+    return !!(this.proc?.isRunning);
```

Comment:
🤯 we can use null coalescing operators now?!

Reply:
this could be configured as babel option ;)